ui-core/src/components/inputs/ComboBox/ComboBox.tsx

```diff
@@ -7,151 +7,139 @@ import React, {
-        ]
-      : []), // Conditionally include the chevron icon only when suggestions are not empty
-  ];
-
   /* eslint-disable react-hooks/exhaustive-deps */
   useEffect(() => {
     if (value) {
-      setInputValue(value);
+      setSelectedOption(value);
+      setInputValue(getSuggestionLabel(value));
     } else {
+      setSelectedOption(undefined);
       setInputValue('');
-      setSelectedOption(null);
     }
   }, [value]);
 
   useEffect(() => {
-    setFilteredSuggestions(sortedSuggestions);
-  }, [sortedSuggestions]);
+    const userInput = normalizeString(inputValue).trim().toLowerCase();
+    if (customOnInputChange) {
+      customOnInputChange(userInput);
+    } else {
+      filterSuggestions(userInput);
+    }
+  }, [inputValue]);
 
-  const handleInputChange: ChangeEventHandler<HTMLInputElement> = (e) => {
-    const userInput = normalizeString(e.currentTarget.value).trim();
-    setInputValue(e.currentTarget.value);
-    onChange?.(e);
+  useEffect(() => {
+    setFilteredSuggestions(suggestions);
+  }, [suggestions]);
 
-    if (userInput.trim() === '') {
-      setFilteredSuggestions([]);
-      setSelectedOption(null);
+  const filterSuggestions = useCallback((input: string) => {
+    if (!input) {
+      setFilteredSuggestions(suggestions);
       return;
     }
 
-    const filtered = sortedSuggestions.filter((suggestion) => {
+    const filtered = suggestions.filter((suggestion) => {
       const suggestionLabel = normalizeString(getSuggestionLabel(suggestion).toLowerCase());
       return exactSearch
-        ? suggestionLabel.startsWith(userInput.toLowerCase())
-        : suggestionLabel.includes(userInput.toLowerCase());
+        ? suggestionLabel.startsWith(input.toLowerCase())
+        : suggestionLabel.includes(input.toLowerCase());
     });
     setFilteredSuggestions(filtered);
+  }, []);
+
+  // behaviors
+  const handleInputChange: ChangeEventHandler<HTMLInputElement> = (e) => {
+    if (selectedOption !== undefined) {
+      setSelectedOption(undefined);
+    }
+
+    const newInputValue = e.currentTarget.value;
+    setInputValue(newInputValue);
   };
 
   const selectSuggestion = (index: number) => {
-    const selectedSuggestion = filteredSuggestions[index];
-    const suggestionLabel = getSuggestionLabel(selectedSuggestion);
-    setInputValue(suggestionLabel);
-    setSelectedOption(selectedSuggestion);
-    onSelectSuggestion?.(selectedSuggestion);
-    setFilteredSuggestions([]);
-    setActiveSuggestionIndex(-1);
-    setTimeout(() => {
-      inputRef.current?.blur();
-    }, 0);
+    const selectedSuggestion = filteredSuggestions.at(index)!;
+
+    onSelectSuggestion(selectedSuggestion);
+    setInputValue(getSuggestionLabel(selectedSuggestion));
+    removeFocus();
+  };
+
+  const closeSuggestions = () => {
+    setIsInputFocused(false);
+
+    // if only 1 suggestion is left, we select it
+    if (filteredSuggestions.length === 1) {
+      selectSuggestion(0);
+      return;
+    }
```

Comment:
That's quite surprising. I'd expect the "Escape" key to cancel everything, instead of selecting the first result. But it seems like the old code was doing that too.

Reply:
Agree, but it was asked by Thibaut I think